### PR TITLE
Refactor slim/html5 templates using Helpers

### PR DESCRIPTION
**Pull request by jirutka**
I’ve refactored Slim templates for HTML5 backend in order to separate most of logic into helpers, make templates more clear, simpler and “DRY”.

This PR includes seven previous (not yet merged) PRs and also some future changes/fixes. I’ll rebase it later to make it more clear, so consider this PR just as a preview for review. The most important commit is [this](https://github.com/jirutka/asciidoctor-backends/commit/d7dce1b40b7e9479ab84253113194f2619780eb9).

I depends on changes in the upcoming Asciidoctor 1.5.2.

### CODE REVIEW

**Comment by mojavelinux** [comment]
> DocTest currently needs Ruby ≥ 2.1.0. I can support 2.0.0 (there’s just one place in the code that isn’t compatible with < 2.1), but I really don’t want to support anything older.

It would be nice to support Ruby 2.0.0 at least until 2.1 is available on Windows. But I'm not going to stress about it.

> Ruby 1.9.3 is obsolete and at the end of life. Since 1.9.3 is at the EOL, it’s IMHO a good time to discuss dropping support of Ruby < 2.0.0 in Asciidoctor core

I plan to drop older Rubies in 1.6.0, but not in the 1.5.x series (just to avoid distractions). We've already said we'll be dropping 1.8.7. I haven't made a decision yet about 1.9.3. I will strongly consider it, but I do want to make sure that we don't leave the mainstream users high and dry (trust me, it becomes a major issue for me when this stuff doesn't just work for them).

For what it's worth, Asciidoctor PDF & EPUB3 require at least 1.9.3 and soon to be 2.0.0 at a minimum because of minimum dependencies of requirements of dependencies.

> the same probably applies for jruby-19mode.

That's a totally separate thing. JRuby 1.7 doesn't have a complete 2.0 mode, so technically jruby-19mode is the most current...and in some ways has partial 2.0 support. Things will clean up once JRuby 9000 is released...but running jruby-head on Travis CI takes _forever_.

Having said that, I do think we should add at least jruby-19mode or jruby-head to Travis CI so we know this stuff is tested for AsciidoctorJ users.

File: slim/html5/block_audio.html.slim [hunk 1]
```diff
@@ -1,6 +1,8 @@
-.audioblock id=@id class=[@style,role]
-  - if title?
-    .title=captioned_title
+= block_with_captitle ['audioblock', style]
```
Comment (jirutka) [inline slim/html5/block_audio.html.slim]: I’ve found that it’s actually unnecessary to explicitly distinct between captioned and uncaptioned titles — method `captioned_title` is [defined in AbstractBlock](https://github.com/asciidoctor/asciidoctor/blob/4d0ae51641c773004355879f5642eb18b6046dce/lib/asciidoctor/abstract_block.rb#L142-L146) and it returns just naked title when the nodes doesn’t have a caption.

So problem solved, we don’t need extra helper for captioned titles at all.

**Comment by jirutka** [comment]
DocTest now supports MRI 2.0.0 and JRuby 9000-dev. Rubinius doesn’t support keyword arguments yet, so it’s out of game for now.

However, there’s some problem with installing Slim on JRuby (see [build log](https://travis-ci.org/asciidoctor/asciidoctor-backends/jobs/45737037)).

btw JRuby is really VERY slooow — both on my machine and Travis — I wonder how anyone can use it…

**Comment by jirutka** [comment]
The problem on JRuby causes [posix-spawn](https://github.com/rtomayko/posix-spawn), dependency of pygments.rb.

Is it really necessary to support pygments.rb when we have CodeRay?

**Comment by jirutka** [comment]
I’ve just found that you already know a lot about issues with pygments.rb. Well, I’ll disable pygments.rb when running on JRuby.

**Comment by jirutka** [comment]
Okay, ready for merge! :+1:

Tests on jruby-head passes except `block_listing:source_highlighter_pygments` (obviously). I’ll add option for selective exclusion of examples in the next version of DocTest.

**Comment by mojavelinux** [comment]
> we don’t need extra helper for captioned titles at all.

\o/

**Comment by mojavelinux** [comment]
> JRuby is really VERY slooow — both on my machine and Travis — I wonder how anyone can use it…

It's definitely not meant for running single executions (like a cli or test suite). The benefit of JRuby is when it runs operations at scale. In that domain, it breezes past anything that doesn't have a hotspot compiler like MRI. It also essential for integration into other JVM languages, hence the value of AsciidoctorJ.

**Comment by mojavelinux** [comment]
> Is it really necessary to support pygments.rb when we have CodeRay?

No, just cut it. In fact, we are strongly considering cutting Pygments from Asciidoctor in favor of [Rouge](https://github.com/jneen/rouge), which is a Pygments port to Ruby. The situation with Pygments.rb is just messy.

**Comment by mojavelinux** [comment]
> Okay, ready for merge! 

Super, I'll do a final review asap, but I'm optimistic we've got it the way we want it!

**Comment by mojavelinux** [comment]
I should be able to get to this in the next day or 2. Finally managed to fend off all the New Year's resolution-fueled e-mails hitting my inbox :)

**Comment by mojavelinux** [comment]
Looks good! It's time to take the plunge! Thanks for your awesome work on this refactoring and test suite.

:coffee: + :beers: + :chocolate_bar:

**Pull request merged**